zerver/tests/test_channel_permissions.py

```diff
@@ -746,3 +748,241 @@ def test_change_stream_message_retention_days_requires_realm_owner(self) -> None
+
+        hamletcharacters_group = NamedUserGroup.objects.get(name="hamletcharacters", realm=realm)
+        moderators_group = NamedUserGroup.objects.get(
+            name=SystemGroups.MODERATORS, realm=realm, is_system_group=True
+        )
+        nobody_group = NamedUserGroup.objects.get(
+            name=SystemGroups.NOBODY, realm=realm, is_system_group=True
+        )
+        members_group = NamedUserGroup.objects.get(
+            name=SystemGroups.MEMBERS, realm=realm, is_system_group=True
+        )
+
+        api_parameter_name_dict = dict(
+            name="new_name",
+            deactivated="is_archived",
+            invite_only="is_private",
+        )
+        api_parameter_name = property_name
+        if property_name in api_parameter_name_dict:
+            api_parameter_name = api_parameter_name_dict[property_name]
+
+        data = {}
+        if property_name in Stream.stream_permission_group_settings:
+            data[api_parameter_name] = orjson.dumps({"new": new_value}).decode()
+        elif not isinstance(new_value, str):
+            data[api_parameter_name] = orjson.dumps(new_value).decode()
+        else:
+            data[api_parameter_name] = new_value
+
+        default_error_msg = "You do not have permission to administer this channel."
+
+        def check_channel_property_update(
+            user: UserProfile, allow_fail: bool = False, error_msg: str = default_error_msg
+        ) -> None:
+            old_value = getattr(stream, property_name)
+            if property_name in Stream.stream_permission_group_settings:
+                if not hasattr(old_value, "named_user_group"):
+                    old_value = get_group_setting_value_for_api(old_value)
+                else:
+                    old_value = old_value.named_user_group
+
+            if property_name == "deactivated" and new_value is True:
+                # There is a separate endpoint for deactivating streams.
+                result = self.api_delete(user, f"/api/v1/streams/{stream.id}")
+            else:
+                result = self.api_patch(user, f"/api/v1/streams/{stream.id}", info=data)
+
+            if allow_fail:
+                self.assert_json_error(result, error_msg)
+                return
+
+            self.assert_json_success(result)
+            stream.refresh_from_db()
+            if property_name in Stream.stream_permission_group_settings:
+                self.assertEqual(getattr(stream, property_name + "_id"), new_value)
+            else:
+                self.assertEqual(getattr(stream, property_name), new_value)
+
+            # Reset to original value
+            if property_name in Stream.stream_permission_group_settings:
+                do_change_stream_group_based_setting(
+                    stream, property_name, old_value, acting_user=iago
+                )
+            else:
+                setattr(stream, property_name, old_value)
+                stream.save(update_fields=[property_name])
```

Comment:
Can the above two if else blocks be combined? Will make it more readable.

Reply:
We need to call `do_change_stream_group_based_setting` to handle anonymous groups, so we cannot call `setattr` directly.